Recommend setting publish options in project file

sbomer opened the pull request:
The `PublishTrimmed` option also disables trim-incompatible features, and turns on a Roslyn analyzer for trim compatibility, so it should be set in the project file to ensure these settings apply during a `dotnet build`.

Similarly, `PublishSingleFile` enables a single-file analyzer so it should also be set in the project file.

Addresses part of https://github.com/mono/linker/issues/2085.

sbomer committed:
Recommend setting PublishTrimmed in project file

sbomer committed:
More fixes for single-file and publishing steps

sbomer committed:
Fix typo in ilnk

LakshanF commented on docs/core/deploying/trimming-options.md:
```diff
@@ -16,10 +16,14 @@ Trimming with `PublishTrimmed` was introduced in .NET Core 3.0. The other option
 
 This will trim any assemblies which have been configured for trimming. With `Microsoft.NET.Sdk` in .NET 6, this includes the any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for framework assemblies. In .NET 5, framework assemblies from the netcoreapp runtime pack are configured for trimming via `<IsTrimmable>` MSBuild metadata. Other SDKs may define different de ...
 
+Starting in .NET 6, this setting also enables the trim compatibility [Roslyn analyzer](#roslyn-analyzer), and disables [features](#trimming-framework-library-features) which are incompatible with trimming.
```
Should we update the features list to include the new ones like BuiltInComInteropSupport, EnableCppCLIHostActivation etc.?

sbomer replied:
I don't think we should document the ones that are already disabled by default when linking - we don't want to encourage people to enable the unsupported ones.

But it looks like `UseNativeHttpHandler` is missing from the list. @eerhardt do you think we should document that one? It looks to me like `UseNativeHttpHandler=true` will prevent some (but not all) references to `HttpClientHandler` - I'm not sure whether this feature was designed for trimming in particular.

eerhardt replied:
I do find it a bit confusing that we are explicitly saying `disables features [link] which are incompatible with trimming`, but then the link doesn't list any features which are incompatible with trimming.

Maybe we should have 2 sections? Features that can be toggled when trimming, and features that are not compatible with trimming?

> But it looks like UseNativeHttpHandler is missing from the list.

@EgorBo @akoeplinger @steveisok @marek-safar - should we be documenting `UseNativeHttpHandler` here? I believe it is only valid for Android and iOS apps.

sbomer replied:
I added  a section for the trim-incompatible features, and also added a few more switches to the optional features section, including UseNativeHttpHandler.

agocke commented on docs/core/tools/dotnet-publish.md:
```diff
@@ -153,13 +153,13 @@ For more information, see the following resources:
-  We recommend that you specify this option in a publish profile rather than on the command line. For more information, see [MSBuild](#msbuild).
+  We recommend that you specify this option in the project file rather than on the command line. Setting it in the project file enables a Roslyn analyzer that provides single-file compatibility warnings during `dotnet build`. For more information, see [MSBuild](#msbuild).
 
 - **`-p:PublishTrimmed=true`**
```
```suggestion
- **`<PublishTrimmed>true</PublishTrimmed>`**
```

agocke replied:
Same for `PublishSingleFile` above

sbomer replied:
Unfortunately this document is talking about options for the `dotnet publish` command. Should we maybe just remove these sections from this file? It's strange that they're documented as publish options when really they're just passed to MSBuild. And these same docs then recommend _not_ using the command-line arg version.

agocke replied:
Removing them seems best, although the unfortunate benefit is that if you specific `PublishTrimmed` on the command line then it applies to all projects, and would turn on the analyzer for all of them. I suppose we also need to recommend adding `EnableTrimmingAnalyzer` in every project.

sbomer replied:
I removed PublishTrimmed, PublishSingleFile, and PublishReadyToRun. @tdykstra does this seem ok to you?

eerhardt commented on docs/core/deploying/single-file.md:
```diff
@@ -139,18 +139,26 @@ These properties have the following functions:
+<PropertyGroup>
+    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
+    <PublishSingleFile>true</PublishSingleFile>
+    <SelfContained>true</SelfContained>
```
You don't need to set `SelfContained=true` explicitly. It is defaulted that way when you set `RuntimeIdentifier=foo`.

https://github.com/dotnet/sdk/blob/c53c8ac81e1bcb8118c31bc4f5622396c72e9959/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets#L107-L115

eerhardt commented on docs/core/deploying/single-file.md:
```diff
@@ -139,18 +139,26 @@ These properties have the following functions:
 Publish a single file application using the [dotnet publish](../tools/dotnet-publish.md) command. When you publish your app, set the following properties:
 
 - Publish for a specific runtime: `-r win-x64`
```
It's a little odd that we are mixing `-r` (the command line option) here and the project setting for `PublishSingleFile` in the next line. I think it would make sense to be consistent here.

eerhardt commented on docs/core/deploying/single-file.md:
````diff
@@ -139,18 +139,26 @@ These properties have the following functions:
 The following example publishes an app for Windows as a self-contained single file application.
 
-```dotnetcli
-dotnet publish -r win-x64 -p:PublishSingleFile=true --self-contained true
````
Maybe we should list both the XML project settings and the command line options, and then explain the advantages to setting the XML project settings?

My concern here is that new developers could be confused that they can't set these options on the command line (especially the `RuntimeIdentifier`).

tdykstra commented on docs/core/deploying/trimming-options.md:
```diff
@@ -16,10 +16,14 @@ Trimming with `PublishTrimmed` was introduced in .NET Core 3.0. The other option
+
+Place this in the project file to ensure that the setting applies during `dotnet build`, not just `dotnet publish`.
 
 This will trim any assemblies which have been configured for trimming. With `Microsoft.NET.Sdk` in .NET 6, this includes the any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for framework assemblies. In .NET 5, framework assemblies from the netcoreapp runtime pack are configured for trimming via `<IsTrimmable>` MSBuild metadata. Other SDKs may define different de ...
```
```suggestion
This will trim any assemblies that have been configured for trimming. With `Microsoft.NET.Sdk` in .NET 6, this includes the any assemblies with `[AssemblyMetadata("IsTrimmable", "True")]`, which is the case for framework assemblies. In .NET 5, framework assemblies from the netcoreapp runtime pack are configured for trimming via `<IsTrimmable>` MSBuild metadata. Other SDKs may define different defaults.
```

tdykstra commented on docs/core/deploying/trimming-options.md:
```diff
@@ -16,10 +16,14 @@ Trimming with `PublishTrimmed` was introduced in .NET Core 3.0. The other option
+
 ## Trimming granularity
 
 The following granularity settings control how aggressively unused IL is discarded. This can be set as a property affecting all trimmer input assemblies, or as metadata on an [individual assembly](#trimmed-assemblies) which overrides the property setting.
```
```suggestion
The following granularity settings control how aggressively unused IL is discarded. This can be set as a property affecting all trimmer input assemblies, or as metadata on an [individual assembly](#trimmed-assemblies), which overrides the property setting.
```

tdykstra reviewed (approved):
LGTM, a couple suggestions

sbomer committed:
Apply suggestions from code review

Co-authored-by: Tom Dykstra <tdykstra@microsoft.com>

sbomer committed:
Document features disabled when trimming

Also document UseNativeHttpHandler, AutoreleasePoolSupport, and MetadataUpdaterSupport

sbomer committed:
Remove publish properties from dotnet-publish docs

sbomer committed:
PR feedback

- Remove unnecessary SelfContained
- Mention project properties and command-line arguments

sbomer committed:
Try to fix link

eerhardt commented on docs/core/deploying/trim-self-contained.md:
```diff
@@ -73,8 +73,8 @@ If we're using .Net 5.0, we can update our project file to include the following
 
 Trim your application using the [dotnet publish](../tools/dotnet-publish.md) command. When you publish your app, set the following properties:
```
Should this line be updated? It makes it sound like I should only change my project when I pubilsh.

```suggestion
Trim your application using the [dotnet publish](../tools/dotnet-publish.md) command. Set the following properties in your project file, then publish the app using `dotnet publish`.
```

eerhardt commented on docs/core/deploying/trim-self-contained.md:
```diff
@@ -73,8 +73,8 @@ If we're using .Net 5.0, we can update our project file to include the following
 
-- Publish as a self-contained app for a specific runtime: `-r win-x64`
-- Enable trimming: `/p:PublishTrimmed=true`
+- Publish as a self-contained app for a specific runtime: `<RuntimeIdentifier>win-x64</RuntimeIdentifier>`
```
Generally, we've been telling people not to hard-code RIDs into their project file, and instead pass them on the command line / publish profile.

cc @dsplaisted 

Are we telling people to put the RID in the project now because without it, `PublishTrimmed` breaks `dotnet build`? If yes, maybe we should move the validation to the `publish` command that ensures you are publishing SelfContained if `PublishTrimmed=true`.

sbomer replied:
No, `PublishTrimmed` doesn't break build - I was just changing this for consistency with the line below to address https://github.com/dotnet/docs/pull/24740#discussion_r655498139.  I'll change it back to recommending `-r win-x64` and `<PublishTrimmed>true</PublishTrimmed>` but being more explicit about what should go in the project file.

eerhardt commented on docs/core/tools/dotnet-publish.md:
```diff
@@ -139,28 +138,6 @@ For more information, see the following resources:
 
     If you specify a relative path when publishing a solution, each project's output goes into a separate folder relative to the project file location. If you specify an absolute path when publishing a solution, all publish output for all projects goes into the specified folder.
 
-- **`-p:PublishReadyToRun=true`**
```
Are we concerned that removing these options here decreases the chances of people finding them and learning about what they do? I think it is discoverable to list these publish options in the "dotnet publish" document, so people discover them.

sbomer replied:
I'm slightly concerned, but removed it based on the discussion in https://github.com/dotnet/docs/pull/24740#discussion_r654626080. We do have dedicated documentation for these properties. Maybe the best thing would be to link to the important publish MSBuild options from the publish options doc, without mentioning them as command-line arguments? cc @agocke

eerhardt replied:
> link to the important publish MSBuild options from the publish options doc, without mentioning them as command-line arguments?

I like that approach. That way they are still discoverable.

eerhardt commented on docs/core/deploying/trimming-options.md:
```diff
@@ -197,8 +205,43 @@ Several feature areas of the framework libraries come with linker directives tha
+
+- `<EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization>`
+
+   [`DesigntimeLicenceContextSerializer`](https://docs.microsoft.com/dotnet/api/system.componentmodel.design.designtimelicensecontextserializer?view=net-5.0) use of BinaryFormatter serialization is dissabeld.
```
```suggestion
   [`DesigntimeLicenceContextSerializer`](https://docs.microsoft.com/dotnet/api/system.componentmodel.design.designtimelicensecontextserializer?view=net-5.0) use of BinaryFormatter serialization is disabled.
```

eerhardt commented on docs/core/deploying/trimming-options.md:
```diff
@@ -197,8 +205,43 @@ Several feature areas of the framework libraries come with linker directives tha
+
+    C++/CLI host activation is disabled.
+
+- `<EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization>`
```
Hmm, it is interesting that `EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization` is disabled by default when you trim, but `EnableUnsafeBinaryFormatterSerialization` isn't.

Should `EnableUnsafeBinaryFormatterSerialization` also be set to `false` by default when you trim?

cc @GrabYourPitchforks @pgovind

eerhardt replied:
I've re-opened https://github.com/dotnet/sdk/issues/14475 for this issue.

sbomer committed:
Apply suggestions from code review

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

sbomer committed:
Rework trim/single-file CLI steps

sbomer committed:
Add back publish MSBuild options for dotnet-publish

In a separate section that mentions MSBuild properties.

eerhardt commented on docs/core/tools/dotnet-publish.md:
```diff
@@ -57,6 +56,30 @@ dotnet publish -p:PublishProfile=FolderProfile
+
+  Compiles application assemblies as ReadyToRun (R2R) format. R2R is a form of ahead-of-time (AOT) compilation. For more information, see [ReadyToRun images](../deploying/ready-to-run.md). Available since .NET Core 3.0 SDK.
+
+  To see warnings about missing dependencies that could cause runtime failures, use `PublishReadyToRunShowWarning`.
```
`PublishReadyToRunShowWarning=true` ?

eerhardt commented on docs/core/tools/dotnet-publish.md:
```diff
@@ -57,6 +56,30 @@ dotnet publish -p:PublishProfile=FolderProfile
+
+  To see warnings about missing dependencies that could cause runtime failures, use `PublishReadyToRunShowWarning`.
+
+  We recommend that you specify this option in a publish profile rather than on the command line.
```
"this option" is a bit ambiguous between `PublishReadyToRun` and `PublishReadyToRunShowWarning` here.

eerhardt commented on docs/core/tools/dotnet-publish.md:
```diff
@@ -57,6 +56,30 @@ dotnet publish -p:PublishProfile=FolderProfile
+
+- `PublishSingleFile`
+
+  Packages the app into a platform-specific single-file executable. The executable is self-extracting and contains all dependencies (including native) that are required to run the app. When the app is first run, the application is extracted to a directory based on the app name and build identifier. Startup is faster when the application is run again. The application doesn't need to extract itself ...
```
> When the app is first run, the application is extracted to a directory based on the app name and build identifier

I know this was already there, but this is no longer the case for 5.0+.

eerhardt reviewed (approved):
Looks great. Thanks for making these changes, @sbomer!

sbomer committed:
PR feedback

- Remove outdated single-file description
- Avoid ambiguous "this option"
- Specify value for PublishReadyToRunShowWarning
